Allow registering "runtime classes" from GDExtension

dsnopek opened the pull request:
Classes registered via GDExtension normally run in the editor (just like module classes) which is sort of similar to `@tool` scripts. This is the number 2 complaint from people who attempt to use GDExtension to write gameplay code (as opposed to people who are using GDExtension to extend the engine).

This PR allows registering "runtime classes" from GDExtension, where the real code doesn't run in the editor, similar to how non-`@tool` scripts work.

The idea is that when running in the editor, these special classes would have "placeholders" created, which can store the classes properties, so that the inspector and saving/loading still works (which is similar to how non-`@tool` scripts are implement too).

~~In order to do this, this PR refactors how `Object` handles extension instances. Previously, this was just a `void *` and everything in `Object` and elsewhere had to interact with the GDExtension interface directly. This PR adds a `ObjectGDExtensionInstance` class to centralize this work, so now we have the `ObjectGDExtension` and `ObjectGDExtensionInstance` classes which are analogous to the `Script` and `ScriptInstance` classes.~~

~~This makes it easy for the `ObjectGDExtensionInstance` class to switch over to "placeholder mode" when necessary.~~

**UPDATE:** I ended up rolling back the refactor, because I came up with a clean way to handle placeholders without it.

For this PR to work with godot-cpp, you need this companion PR https://github.com/godotengine/godot-cpp/pull/1256

### Still TODO

- ~~Get the default values for the real classes properties (we may need to create a single instance the real class to do this - I think this should be possible to do)~~
- ~~Set/get property values and ensure saving and loading works~~
- ~~Correctly handle reverting properties via the inspector~~
- ~~See if `ObjectGDExtensionInterface` can do something smart for the `GDVIRTUAL*()` macros~~
- ~~Ensure that this works with hot reload (it should)~~
- ~~Investigate the interaction between runtime classes and tool GDScripts~~
- ~~Investigate the interaction between runtime and non-runtime classes from the same GDExtension~~

Fixes https://github.com/godotengine/godot/issues/54999

_Production edit: closes godotengine/godot-roadmap#36_

dsnopek commented:
Actually, this seems to be mostly working? I haven't managed to get classes switching back and forth between "gameplay" and "not gameplay" on a hot reload, but otherwise the basics are working.

The implementation feels kinda messy, and I'm not sure if this is the right approach in general, so this'll need some more testing (to see if it meets expectations) and discussion.

dsnopek commented:
I've been thinking about this PR over the weekend. I think what it's doing at a high-level is probably the right overall approach (ie. making placeholders when gameplay classes are instantiated in the editor). However, the implementation is kinda messy, and I'm not crazy about it.

Perhaps the way that `Object` interacts with its extension and extension instance needs to be refactored? For scripts, we have the `Script` and `ScriptInstance` classes, and the way `Object` interacts with them is pretty straight-forward. However, for extensions, we have the `ObjectGDExtension` _struct_ and a `void *` for the instance, and all the interactions are done manually all over `Object`.

Perhaps if we wrapped at least the extension instance in a light-weight class, the interactions could be made cleaner and easier to maintain? Since performance is important in GDExtension, we'd probably want to make sure that all the hot-path functions are `_FORCE_INLINE_` to try and keep the compiled code the same. But that would also make it much easier to optionally create a placeholder version of an extension object (and, of course, that additional code could be made to only exist in editor builds).

In any case, I may experiment with some ways to refactor how `Object` interacts with its extension and extension instance before continuing this PR.

Bromeon commented:
As prior art, I worked around this issue in godot-rust: https://github.com/godot-rust/gdext/pull/365

Basically, what you call gameplay classes is the default, and a `#[class(tool)]` attribute can be used to make classes run in the editor. "Not running" amounts to not having virtual callbacks `_ready`, `_process`, `_notification` etc invoked, but classes are still constructed and accessible (this is necessary because scene trees break otherwise).

I haven't built a larger game with this yet, so there's some research to be done how useful that behavior is in practice. But so far, community feedback has been quite positive.

bluenote10 commented:
> In the implementation in this PR, we don't actually create the real class (except for once, in order to get the property defaults)

Would this still allow for preventing `_init` to be called whatsoever?

Use case: I often want to load certain external resources already at construction time, so that the class can consider "all external resources are loaded" as an invariant, i.e., not having to deal with partial initialization. If I remember correctly 
https://github.com/godot-rust/gdext/pull/365 only prevents calling virtual methods like `_ready`, `_process`, etc., but not `_init`, which means that when starting the Godot editor I see all my gdextension classes doing their initialization. Some of them perform some rather non-trivial tasks sometimes even with side effects (e.g. logger setup, singleton initialization, file system discovery...). These nodes/objects are internal and are not supposed to be used directly, so it would be great if construction by the editor could be prevented entirely.

dsnopek commented:
> Would this still allow for preventing `_init` to be called whatsoever?

Yes.

With the current implementation in this PR, each "real" class needs to be created exactly once in order to get the default values for its properties, but it's destroyed right after that, and all subsequent instances of the class are just "placeholders". The placeholders don't create an instance of your real class, they just have a hash map of the properties, so they can be edited in the inspector.

Bromeon commented:
> With the current implementation in this PR, each "real" class needs to be created exactly once in order to get the default values for its properties

Question regarding that, what if the class cannot be constructed (GDExtension doesn't provide a `create` callback, or marks it abstract)?

dsnopek commented:
> Question regarding that, what if the class cannot be constructed (GDExtension doesn't provide a `create` callback, or marks it abstract)?

Hm. I think the current PR doesn't correctly handle that situation, but I think what it _should_ do, is also not allow creating a placeholder if the class can't normally be created. I'll add a fix for that when I come to working on this one. :-)

dsnopek committed:
Allow registering "runtime classes"

dsnopek commented:
I've finally finished my big refactor of how `Object` handles extension instances, mostly as described in [my earlier comment](https://github.com/godotengine/godot/pull/82554#issuecomment-1743081835).

Previously, `Object` just held a `void *` for the extension instance we got from the binding and everything in `Object` (and elsewhere) had to interact with the GDExtension interface directly. This PR adds a `ObjectGDExtensionInstance` class to centralize this work, so now we have the `ObjectGDExtension` and `ObjectGDExtensionInstance` classes which are analogous to the `Script` and `ScriptInstance` classes.

This makes it easy for the `ObjectGDExtensionInstance` class to switch over to "placeholder mode" when necessary.

The main difference between what I wrote in my comment and what I implement has to do with this:

> Since performance is important in GDExtension, we'd probably want to make sure that all the hot-path functions are `_FORCE_INLINE_` to try and keep the compiled code the same as it is now.

I originally did try using `_FORCE_INLINE_` on everything, but the functions were just so long and so ugly, it felt dirty having them in `object.h` which most of Godot includes. So, they are normal functions in this PR currently, which does technically add another function call which could theoretically hurt performance, but I guess I'd argue that the existing functions aren't strictly "hot-pathes". Maybe `set()` and `get()`? I think the `GDVIRTUAL*()` macros are actual "hot-pathes" and those remain unconverted for now (although, I may do a thing with `_FORCE_INLINE_`, that's one of my remaining todo's.)

There's only a handful of things left to do on this (see the PR description) before it's ready for review!

dsnopek commented:
I just pushed the following changes:

- Moved `set()`, `get()` and `notification()` to the header with `_FORCE_INLINE_` since I think those could arguably be called "hot-pathes"
- Also added `get_virtual()` and `call_virtual()` (also `_FORCE_INLINE_`) to be used in the `GDVIRTUAL*()` macros
- Various little renames and clean-ups

I also tested using hot-reload with a gameplay class, and it seemed to work fine!

Assuming tests pass, this should finally be ready for review :-)

dsnopek commented:
Actually, thinking about it some more... I think I know a way that I could do this without the refactor into `ObjectGDExtensionInstance`, but that isn't as messy as the older version. I guess it's a question of: Do we prefer to centralize the concept of an "extension instance" into a class? Or, have it scattered across the code base -- which is basically inlining, and consequently, good for performance?

dsnopek commented:
@Ansraer brought up some interesting edge cases on RocketChat that I haven't accounted for:

1. **Interaction between gameplay classes and tool GDScripts:** With GDScript, if a tool GDScript creates an instance of a non-tool GDScript, it creates a real instance of the class. Whereas in my current PR, any time you create an instance of a gameplay class, you get the placeholder version. Users will probably expect the same behavior as GDScript. To do this, I think we need to only create the placeholders when loading a scene, but create the real version whenever instantiated directly
2. **Interaction between gameplay and non-gameplay classes from the same GDExtension:** If you have a scene whose root is a non-gameplay class, and it contains a child node that is a gameplay class, what do you get if you `get_node()` that child node? When run in game, you'd always get a real instance of the gameplay class, but in the editor, that would be a placeholder, which I think would be the type of the closest native parent rather than the type defined in the GDExtension. I think GDScript should have a different version of the same problem with tool and non-tool scripts, so it'd be interesting to see how it handles it.

Rennan24 commented:
@dsnopek 
> 2. **Interaction between gameplay and non-gameplay classes from the same GDExtension:** If you have a scene whose root is a non-gameplay class, and it contains a child node that is a gameplay class, what do you get if you `get_node()` that child node? When run in game, you'd always get a real instance of the gameplay class, but in the editor, that would be a placeholder, which I think would be the type of the closest native parent rather than the type defined in the GDExtension. I think GDScript should have a different version of the same problem with tool and non-tool scripts, so it'd be interesting to see how it handles it.

In C# at least when you create a Tool class and try to reference or GetNode() a Non-Tool class, it creates an InvalidCastException because there is no actual Non-Tool class running in the editor. Although in GDScript it works perfectly fine which is a little weird to me but I don't know the inner workings of the C# or GDScript bindings.

There are multiple people that have run into this problem and its something that is pretty annoying because you then have to mark everything as [Tool] or just use the inherited class which loses some information about the class
(#80298, #40069, #85825)


I attached a MRP showing an example of what I mean if I didn't explain it properly! don't know if this necessarily applies to this new gameplay classes but worth bringing up :D
[CSvGD Tool Scripts (MRP).zip](https://github.com/godotengine/godot/files/13845476/CSvGD.Tool.Scripts.MRP.zip)

Hopefully its something that can be fixed in GDExtensions or C# since my preferred use case is to have a [Tool] script that can automatically get references with GetNode() or just perform some operations on its children that are NonTool classes

dsnopek commented:
I just done another almost complete rewrite of this PR! :-)

I've reverted the refactor in my last big rewrite. I did quite like it, however, I came up with a clean way to inject the placeholders with the refactor (much better than the mess that was in my first iteration), and I figure smaller changes are better (this new iteration adds ~200 fewer lines), and this eliminates the the "inlining" question I debated with myself above.

I've also "solved" the cases brought up by @Ansraer, meaning I've tested how these cases work with GDScript and C#, made it so that GDExtension works in a roughly comparable way, and did a bunch of testing.

1. Re: **The interaction between gameplay classes and tool GDScripts.**
    - A tool GDScript can grab a node for a gameplay class from the scene (which would be a placeholder) and get/set its properties, but calling any of its methods is a no-op. This is comparable to accessing a non-tool GDScript, except in that case you can't even access its properties.
    - When a tool GDScript directly creates an instance of gameplay class, however, a real instance will be created. This matches what happens when a tool GDScript creates an instance of a non-tool GDScript.
2. Re: **Interaction between gameplay and non-gameplay classes from the same GDExtension.**
    - When a non-gameplay class is the root node of a scene, and it accesses a child node for a gameplay class from the scene (which would be a placeholder), it will get an instance of the closest native parent class. This isn't ideal, because you'll get a different type in game versus in the editor, it but matches what happens with C# currently. (Note: if you use `get()` or `set()` you'll still be able to access the custom properties from the gameplay class.)
    - But when a non-gameplay class directly creates an instance of a gameplay class, it'll get a real instance. This matches GDScripts behavior and the behavior in point nr 1.

@Rennan24 Thanks for all the references and the comprehensive explanation of how this is currently working with GDScript and C#! This helped a lot when starting my own investigation and experimentation. :heart: 

> Hopefully its something that can be fixed in GDExtensions or C# since my preferred use case is to have a [Tool] script that can automatically get references with GetNode() or just perform some operations on its children that are NonTool classes

Unfortunately, I'm not sure we can do much better. The whole point of gameplay classes (or non-tool scripts), is that we don't create a real instance of them in the editor. So, when a normal GDExtension class or tool script grabs the placeholder, there's no way to give them a real instance, because we just don't have one. I think the best we can do is allow access to its custom properties, and allow it to have the behavior of its native parent class.

dsnopek commented:
I did some more testing with hot reload and found it wasn't cleaning up properly. This should be fixed in my latest push!

allenwp commented:
Is the term "gameplay" used anywhere else in the codebase/interface to describe a non-`tool` class/script? Or is this a new term that is being introduced by this PR?

allenwp commented:
Is it possible to register as an abstract class to create new `Node` types in a GDExtension that cannot be instantiated in the scene? For example, a GDExtension library would need to be able to create "gameplay" classes and "abstract gameplay" classes to create both types of `Nodes`.
![image](https://github.com/godotengine/godot/assets/17506573/55b8ca74-0a84-4ac4-80ed-edce45069376)

dsnopek commented:
> Is the term "gameplay" used anywhere else in the codebase/interface to describe a non-`tool` class/script? Or is this a new term that is being introduced by this PR?

The term is invented here. Since we don't actually have `tool` classes (it's just the default for engine classes), I didn't like calling them non-`tool` classes, and the best I could think of was "gameplay".

If you have any better ideas for terms, please let me know!

> Is it possible to register as an abstract class to create new Node types in a GDExtension that cannot be instantiated in the scene? For example, a GDExtension library would need to be able to create "gameplay" classes and "abstract gameplay" classes to create both types of Nodes.

You could make an "abstract gameplay" class (with the API directly - I didn't add a function for it in my godot-cpp changes), but there isn't much point, because you can't instantiate an abstract class, and gameplay classes are only special in that instantiating them in the editor creates a placeholder. So, an "abstract gameplay" class and a "plain old abstract" class will work exactly the same. :-)

So, if you wanted to have an abstract class as the parent of one or more gameplay classes, you can just make a normal abstract class.

allenwp commented:
My gut tells me that "runtime only class" is a better fit than "gameplay class".

This would address (possible?) issues with "gameplay class":
- I would not consider a new `Node` class that I'm creating that reports analytics back to the developer to be anything relating to "gameplay", but it is definitely something that should be only execute in the exported executable's runtime.
- Godot can be used to create non-game apps, so GDExtensions may be made for non-gameplay purposes.

So something that doesn't relate to "gameplay", but instead hints at "will only execute in the exported executable's runtime" would be a better fit. Is there any other _existing_ terminology that describes the exported executable that would make sense here?

allenwp commented:
Hmm. I'm sometimes getting this crash when I add a new `Node` that is registered using the new `register_gameplay_class` function.

```
Exception thrown: read access violation.
this->**_data** was 0x7.
```

```
================================================================
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.3.dev.custom_build (86e169442440131c04354d86bc3c77754443e224)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[0] godot::List<godot::PropertyInfo,godot::DefaultAllocator>::is_empty (C:\Personal\Godot\godot-vector-display\godot-cp)[1] godot::List<godot::PropertyInfo,godot::DefaultAllocator>::is_empty (C:\Personal\Godot\godot-vector-display\godot-cp)[2] vector_display::VDShape3D::get_property_list_bind (C:\Personal\Godot\godot-vector-display\src\vd_shape_3d.h:9)
[3] Object::get_property_list (C:\Personal\Godot\godot_src\core\object\object.cpp:496)
[4] EditorData::instantiate_object_properties (C:\Personal\Godot\godot_src\editor\editor_data.cpp:586)
[5] CreateDialog::instantiate_selected (C:\Personal\Godot\godot_src\editor\create_dialog.cpp:568)
[6] SceneTreeDock::_do_create (C:\Personal\Godot\godot_src\editor\scene_tree_dock.cpp:2575)
[7] SceneTreeDock::_create (C:\Personal\Godot\godot_src\editor\scene_tree_dock.cpp:2647)
[8] call_with_variant_args_helper<SceneTreeDock> (C:\Personal\Godot\godot_src\core\variant\binder_common.h:308)
[9] call_with_variant_args<SceneTreeDock> (C:\Personal\Godot\godot_src\core\variant\binder_common.h:418)
[10] CallableCustomMethodPointer<SceneTreeDock>::call (C:\Personal\Godot\godot_src\core\object\callable_method_pointer.)[11] Callable::callp (C:\Personal\Godot\godot_src\core\variant\callable.cpp:57)
[12] Object::emit_signalp (C:\Personal\Godot\godot_src\core\object\object.cpp:1128)
[13] Node::emit_signalp (C:\Personal\Godot\godot_src\scene\main\node.cpp:3676)
[14] Object::emit_signal<> (C:\Personal\Godot\godot_src\core\object\object.h:927)
[15] CreateDialog::_confirmed (C:\Personal\Godot\godot_src\editor\create_dialog.cpp:434)
[16] call_with_variant_args_helper<CreateDialog> (C:\Personal\Godot\godot_src\core\variant\binder_common.h:308)
[17] call_with_variant_args<CreateDialog> (C:\Personal\Godot\godot_src\core\variant\binder_common.h:418)
[18] CallableCustomMethodPointer<CreateDialog>::call (C:\Personal\Godot\godot_src\core\object\callable_method_pointer.h)[19] Callable::callp (C:\Personal\Godot\godot_src\core\variant\callable.cpp:57)
[20] Object::emit_signalp (C:\Personal\Godot\godot_src\core\object\object.cpp:1128)
[21] Node::emit_signalp (C:\Personal\Godot\godot_src\scene\main\node.cpp:3676)
[22] Object::emit_signal<> (C:\Personal\Godot\godot_src\core\object\object.h:927)
[23] AcceptDialog::_ok_pressed (C:\Personal\Godot\godot_src\scene\gui\dialogs.cpp:122)
[24] AcceptDialog::_text_submitted (C:\Personal\Godot\godot_src\scene\gui\dialogs.cpp:114)
[25] call_with_variant_args_helper<AcceptDialog,String const &,0> (C:\Personal\Godot\godot_src\core\variant\binder_comm)[26] call_with_variant_args<AcceptDialog,String const &> (C:\Personal\Godot\godot_src\core\variant\binder_common.h:418)
[27] CallableCustomMethodPointer<AcceptDialog,String const &>::call (C:\Personal\Godot\godot_src\core\object\callable_m)[28] Callable::callp (C:\Personal\Godot\godot_src\core\variant\callable.cpp:57)
[29] Object::emit_signalp (C:\Personal\Godot\godot_src\core\object\object.cpp:1128)
[30] Node::emit_signalp (C:\Personal\Godot\godot_src\scene\main\node.cpp:3676)
[31] Object::emit_signal<String> (C:\Personal\Godot\godot_src\core\object\object.h:927)
[32] LineEdit::gui_input (C:\Personal\Godot\godot_src\scene\gui\line_edit.cpp:479)
[33] Control::_call_gui_input (C:\Personal\Godot\godot_src\scene\gui\control.cpp:1816)
[34] Viewport::_gui_input_event (C:\Personal\Godot\godot_src\scene\main\viewport.cpp:2235)
[35] Viewport::push_input (C:\Personal\Godot\godot_src\scene\main\viewport.cpp:3351)
[36] Window::_window_input (C:\Personal\Godot\godot_src\scene\main\window.cpp:1616)
[37] call_with_variant_args_helper<Window,Ref<InputEvent> const &,0> (C:\Personal\Godot\godot_src\core\variant\binder_c)[38] call_with_variant_args<Window,Ref<InputEvent> const &> (C:\Personal\Godot\godot_src\core\variant\binder_common.h:4)[39] CallableCustomMethodPointer<Window,Ref<InputEvent> const &>::call (C:\Personal\Godot\godot_src\core\object\callabl)[40] Callable::callp (C:\Personal\Godot\godot_src\core\variant\callable.cpp:57)
[41] Callable::call<Ref<InputEvent> > (C:\Personal\Godot\godot_src\core\variant\variant.h:863)
[42] DisplayServerWindows::_dispatch_input_event (C:\Personal\Godot\godot_src\platform\windows\display_server_windows.c)[43] DisplayServerWindows::_dispatch_input_events (C:\Personal\Godot\godot_src\platform\windows\display_server_windows.)[44] Input::_parse_input_event_impl (C:\Personal\Godot\godot_src\core\input\input.cpp:760)
[45] Input::flush_buffered_events (C:\Personal\Godot\godot_src\core\input\input.cpp:1026)
[46] DisplayServerWindows::process_events (C:\Personal\Godot\godot_src\platform\windows\display_server_windows.cpp:2437)[47] OS_Windows::run (C:\Personal\Godot\godot_src\platform\windows\os_windows.cpp:1476)
[48] widechar_main (C:\Personal\Godot\godot_src\platform\windows\godot_windows.cpp:182)
[49] _main (C:\Personal\Godot\godot_src\platform\windows\godot_windows.cpp:204)
[50] main (C:\Personal\Godot\godot_src\platform\windows\godot_windows.cpp:218)
[51] WinMain (C:\Personal\Godot\godot_src\platform\windows\godot_windows.cpp:232)
[52] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[53] <couldn't map PC to fn name>
-- END OF BACKTRACE --
================================================================
```

I don't seem to get this crash when I register with `register_class`. I can look into this a little more if this doesn't seem entirely obvious. I can also whip up a minimum reproduction sometime later as well.

allenwp commented:
> > Is it possible to register as an abstract class to create new Node types in a GDExtension that cannot be instantiated in the scene? For example, a GDExtension library would need to be able to create "gameplay" classes and "abstract gameplay" classes to create both types of Nodes.
> 
> You could make an "abstract gameplay" class (with the API directly - I didn't add a function for it in my godot-cpp changes), but there isn't much point, because you can't instantiate an abstract class, and gameplay classes are only special in that instantiating them in the editor creates a placeholder. So, an "abstract gameplay" class and a "plain old abstract" class will work exactly the same. :-)
> 
> So, if you wanted to have an abstract class as the parent of one or more gameplay classes, you can just make a normal abstract class.

Hah! You're right -- I didn't think this one through, but I can't see any reason you'd ever need an abstract gameplay class.

dsnopek commented:
@allenwp:

> **EDIT:** It seems that I get this crash when I have two classes and the one that is registered as a gameplay class inherits from the other class. This happens with GDExtension and I haven't tried it in the main codebase (does `register_gameplay_class` work in the main codebase?)
> 
> Here is an example that causes the crash. It's not a full GDExtension project, but adding these files to an existing GDExtension and then trying to add the "VDCircle3D" node to a scene will crash the editor with the above stack trace.

Thanks for testing and digging into this issue! The information about reproducing it was super helpful. :-)

This bug should be fixed in my latest push.

> does `register_gameplay_class` work in the main codebase?

Not yet!

After this PR is merged, I'm planning to make a follow-up PR that allows registering gameplay classes in Godot modules too, in order to remain consistent with godot-cpp's design goal of being API compatible with Godot. And, theoretically, there may be a platform that doesn't support GDExtension, so it'd make it easier to convert a GDExtension to a module, if necessary.

However, I don't know what the other maintainers and contributors will think about that, so I've kept it out of this PR, since I think there is a general consensus around adding this feature for GDExtensions, but adding it for modules still needs to be discussed.

> My gut tells me that "runtime only class" is a better fit than "gameplay class".

I'm not crazy about the "only" in "runtime only class".

You can still create real instances of the class in the editor, but you have to do so explicitly from a `@tool` GDScript/C# or a normal GDExtension class. You only get a placeholder instance when it's created via the editor UI or loading a resource (for example, if you instantiate a `PackedScene`).

A more accurate name would be "runtime only by default class" or something a little too wordy :-)

> Godot can be used to create non-game apps, so GDExtensions may be made for non-gameplay purposes.

While Godot is certainly used to make non-game apps, the developers using it to do so know that it's a game engine, so, personally, I don't think using game-related terminology will be confusing in that case.

allenwp commented:
Thanks for the response!

> While Godot is certainly used to make non-game apps, the developers using it to do so know that it's a game engine, so, personally, I don't think using game-related terminology will be confusing in that case.

Understood. I think calling it a “game [executable only by default] class” would be an improvement if “runtime [executable only by default] class” is not suitable.

I personally interpret the term “gameplay [code]” to be only a part of a “game executable”, which is why I highlight the proposed “gameplay” term as something that could use improvement.

I hope that description makes sense 😅

allenwp commented:
The issue I described previously seems to be fixed on my end! Though, I am getting this crash when I change back and forth from `ClassDB::register_class` and `ClassDB::register_gameplay_class` with a `reloadable = true` GDExtension. I can look into this more later, but figured I'd post the call stack in case that's immediately helpful:

================================================================
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.3.dev.custom_build (18750d709d01e9ce9994183e447b9509dc8257e0)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[0] Object::clear_internal_extension (C:\Personal\Godot\godot_src\core\object\object.cpp:1939)
[1] GDExtension::_clear_extension (C:\Personal\Godot\godot_src\core\extension\gdextension.cpp:1111)
[2] GDExtension::_unregister_extension_class (C:\Personal\Godot\godot_src\core\extension\gdextension.cpp:680)
[3] <couldn't map PC to fn name>
[4] <couldn't map PC to fn name>
[5] GDExtension::deinitialize_library (C:\Personal\Godot\godot_src\core\extension\gdextension.cpp:836)
[6] GDExtensionManager::_unload_extension_internal (C:\Personal\Godot\godot_src\core\extension\gdextension_manager.cpp:)
[7] GDExtensionManager::reload_extension (C:\Personal\Godot\godot_src\core\extension\gdextension_manager.cpp:109)
[8] GDExtensionManager::reload_extensions (C:\Personal\Godot\godot_src\core\extension\gdextension_manager.cpp:251)
[9] call_with_variant_args_helper<GDExtensionManager> (C:\Personal\Godot\godot_src\core\variant\binder_common.h:308)
[10] call_with_variant_args<GDExtensionManager> (C:\Personal\Godot\godot_src\core\variant\binder_common.h:418)
[11] CallableCustomMethodPointer<GDExtensionManager>::call (C:\Personal\Godot\godot_src\core\object\callable_method_poi)
[12] Callable::callp (C:\Personal\Godot\godot_src\core\variant\callable.cpp:57)
[13] CallQueue::_call_function (C:\Personal\Godot\godot_src\core\object\message_queue.cpp:222)
[14] CallQueue::flush (C:\Personal\Godot\godot_src\core\object\message_queue.cpp:328)
[15] SceneTree::physics_process (C:\Personal\Godot\godot_src\scene\main\scene_tree.cpp:473)
[16] Main::iteration (C:\Personal\Godot\godot_src\main\main.cpp:3752)
[17] OS_Windows::run (C:\Personal\Godot\godot_src\platform\windows\os_windows.cpp:1476)
[18] widechar_main (C:\Personal\Godot\godot_src\platform\windows\godot_windows.cpp:182)
[19] _main (C:\Personal\Godot\godot_src\platform\windows\godot_windows.cpp:204)
[20] main (C:\Personal\Godot\godot_src\platform\windows\godot_windows.cpp:218)
[21] WinMain (C:\Personal\Godot\godot_src\platform\windows\godot_windows.cpp:232)
[22] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[23] <couldn't map PC to fn name>
-- END OF BACKTRACE --
================================================================

dsnopek commented:
> Though, I am getting this crash when I change back and forth from `ClassDB::register_class` and `ClassDB::register_gameplay_class` with a `reloadable = true` GDExtension.

Thanks for testing! This crash should be fixed in my latest push :-)

Since this PR started, it's not been allowed to switch to/from gameplay on reload (there's an error message and that class doesn't reload), however, with the latest refactor, I think there may actually be a way! It's not immediately working, but I'll keep poking at it...

dsnopek commented:
> Since this PR started, it's not been allowed to switch to/from gameplay on reload (there's an error message and that class doesn't reload), however, with the latest refactor, I think there may actually be a way! It's not immediately working, but I'll keep poking at it...

I spent a bunch of time working on this, and, unfortunately, I don't think this is actually possible to fully do (or, rather, it's possible in one direction but not the other) because we don't have the full context of the creation of an instance at the point we are reloading.

However, in my latest push I have made it behave a little friendlier when encountering this situation! Rather than failing to reload a class that switches to/from gameplay, it will now do the reload, but just keep its `is_gameplay` flag at whatever value it was at previously.

allenwp commented:
> However, in my latest push I have made it behave a little friendlier when encountering this situation! Rather than failing to reload a class that switches to/from gameplay, it will now do the reload, but just keep its `is_gameplay` flag at whatever value it was at previously.

I just did a very quick test on this and found that the editor no longer crashes and the following error message is reported instead:

`core\extension\gdextension.cpp:446 - GDExtension class 'Example' cannot change to/from gameplay class on hot reload. Restart Godot for this change to take effect.`

Thanks!

allenwp commented:
I'd be curious to hear others thoughts on the terminology, but after a few days I still believe that `register_game_class` would be an improvement over the term `register_gameplay_class`.

The reason being that a "game" consists of:

- gameplay classes
- analytics classes
- achievements classes
- menu classes
- etc.

In my work on game dev teams, "gameplay" is only a part of what goes into a game's executable when referring to classes. This PR is relevant to all of these game classes, not just gameplay classes.

AThousandShips commented on core/extension/gdextension_interface.h:
```diff
@@ -2469,6 +2506,21 @@ typedef void (*GDExtensionInterfaceClassdbRegisterExtensionClass)(GDExtensionCla
+ * @param p_parent_class_name A pointer to a StringName with the parent class name.
+ * @param p_extension_funcs A pointer to a GDExtensionClassCreationInfo2 struct.
+ */
+typedef void (*GDExtensionInterfaceClassdbRegisterExtensionClass3)(GDExtensionClassLibraryPtr p_library, GDExtensionConstStringNamePtr p_class_name, GDExtensionConstStringNamePtr p_parent_class_name, const GDExtensionClassCreationInfo3 *p_extension_funcs);
```
Should deprecate the second version and point the first version to this method

dsnopek replied:
Thanks! Fixed in my latest push

AThousandShips reviewed (approved):
The code looks good, makes sense, haven't tested ATM though

reduz commented on core/object/object.cpp:
```diff
@@ -1791,6 +1799,16 @@ uint32_t Object::get_edited_version() const {
+	// If this is the library this extension comes from and it's a placeholder, we
+	// have to return the closest native parent's class name, so that it doesn't try to
+	// use this like the real object.
+	if (_extension && _extension->library == p_library && _extension->is_placeholder) {
```
would still wrap in `unlikely()`

dsnopek replied:
Thanks! Done in my latest push

reduz commented on core/object/class_db.h:
```diff
@@ -263,7 +268,7 @@ class ClassDB {
 	static bool can_instantiate(const StringName &p_class);
 	static bool is_virtual(const StringName &p_class);
-	static Object *instantiate(const StringName &p_class);
+	static Object *instantiate(const StringName &p_class, bool p_require_real_class = false);
```
I would try to leave the instantiate function clean, and instead add another one like `instantiate_no_extension` or something, so the code is easier to understand.

dsnopek replied:
Ok! I've make it so there's `ClassDB::instantiate(const StringName &p_class)` and `ClassDB::instantiate_no_placeholders(const StringName &p_class)` in my latest push

reduz commented on core/object/object.cpp:
```diff
@@ -492,14 +492,22 @@ void Object::get_property_list(List<PropertyInfo> *p_list, bool p_reversed) cons
+#ifdef TOOLS_ENABLED
+				// If this is a placeholder, we can't call into the GDExtension on the parent class,
+				// because we don't have a real instance of the class to give it.
+				if (!_extension->is_placeholder) {
```
`unlikely()` recommended here too

dsnopek replied:
I think you probably mean `likely()` in this case? This is checking that it's _not_ a placeholder, meaning this branch would be taken by the non-placeholder case. Added as `likely()` in my latest push

reduz reviewed (dismissed):
In general it looks good, save for missing `unlikely()` or `likely()` to help the optimizer. I marked a few but in the more critical functions it would be good to ensure they are there. 
I also feel there is some overlap with MissingResource and MissingNode, but I am not sure if this is fine or intended (I guess its not exactly the same thing).

Other than that it looks great!

reduz commented:
So, here is a thought about this.

If you are extending the engine and using GDExtension, then from your perspective anything that runs in the game will be gameplay.

But if you are using a GDExtension language (say eventually C# or Rust), you will most likely _not_ want to extend the engine and us it all for gameplay.

As such, I am not entirely convinced that this is something we should expose as "gameplay" rather than "tool". At least, as example, I am pretty sure that when C# is moved to GDExtension, it will still want to retain the "tool" concept (given again 99% of the people uses it to create gameplay), so it may want to do this in a flipped up way. 

Has this been discussed?

dsnopek commented:
@reduz Thanks for the review! All your code comments should be addressed in my latest push :-)

> I also feel there is some overlap with MissingResource and MissingNode, but I am not sure if this is fine or intended (I guess its not exactly the same thing).

It's not really the same thing.

There's some high-level similarities in the implementation, but with placeholders for "gameplay classes" (at least if we want to match the behavior of non-`@tool` classes from GDScript) we want them to act like the parent class in the editor (so, if you sub-class `Area3D`, for example, you want it to have all the signals, properties, configuration warnings, gizmo, etc of `Area3D` and its ancestors) and only allow the class's actual custom properties as valid (we instantiate a real version of the class once in the editor to get the property list and all the default values).

> As such, I am not entirely convinced that this is something we should expose as "gameplay" rather than "tool". At least, as example, I am pretty sure that when C# is moved to GDExtension, it will still want to retain the "tool" concept (given again 99% of the people uses it to create gameplay), so it may want to do this in a flipped up way.

We could still use "tool" for the naming, if we said all classes are "tool classes" by default, and you're opting out of being a "tool class" (and I guess call these "non-tool classes"?). However, I personally feel it's kinda awkward when it's the "non-tool classes" that are the special case, especially in verbal/written discussions and in godot-cpp (we already have `ClassDB::register_class()`, so I guess we'd need to have `ClassDB::register_nontool_class()`?).

So, personally, I still prefer using the "gameplay classes" naming, however, I'd be happy to go with whatever alternative gets the community/contributor consensus.

dsnopek commented:
> @dsnopek at this point I wonder, why not making the tool stuff the _non-default_ route? That's probably what reduz is proposing.

Hm, well, even if we change the default for classes registered from GDExtension, we still can't really call them "tool classes" and "normal classes" (at least without some confusion) because "normal classes" means something different (in fact, the complete opposite) for classes registered in the engine. From a naming perspective I think we'd still call them "tool classes" and "nontool classes", or at least I would. :-) So, to me, the default is a separate issue - unless we're talking about changing the default for classes registered in engine too?

dsnopek commented:
It seems like the primary remaining issue is what to name the different types of classes :-)

These seem to be the most popular options based on previous discussions here and on RocketChat:

1. **"Gameplay classes"** - the main downsides are that this invents a new term and may be confusing given that Godot can create apps, or that a "gameplay class" may not concern actual gameplay. The main upside, is that it's a term, not an attempt at a literal description of how the class works (which would likely be an incomplete description, and less memorable).
2. **"Runtime classes"** - the main downside is that this is a more generic description-y term. The main upside is that it removes any potential confusion related to making non-games or using these classes for non-gameplay.
3. **"Tool vs non-tool classes"** - the main upside is that this is an existing term, and the main downside is "non-tool classes" are the special case, and referring to them with a negative feels sort of awkward.

I'll ask again on RocketChat too...

dsnopek commented:
Based on an informal poll on RocketChat, it appears that "runtime class" is the most popular name. So, in my latest push, I've re-named everything. Please let me know what you think!

deralmas reviewed (commented):
I looked at this slightly harder than usual and, outside of a small nitpick, this actually looks fine! :D

Also, as this part of the engine is one I'd like to know a bit better, I made a bunch of questions regarding how this works, as there are some details that I'm missing. Hope you don't mind.

dsnopek commented on core/extension/gdextension.cpp:
```diff
@@ -434,6 +473,9 @@ void GDExtension::_register_extension_class_internal(GDExtensionClassLibraryPtr
 	extension->gdextension.is_abstract = p_extension_funcs->is_abstract;
 	extension->gdextension.is_exposed = p_extension_funcs->is_exposed;
+#ifdef TOOLS_ENABLED
+	extension->gdextension.is_runtime = is_runtime;
```
> Do we really need to but this assignment in a conditional block?

Well, I tried to make it so that this feature only exists in the editor and doesn't add anything to the export templates. I don't know that it _needs_ to be done that way, but I personally prefer it, because the compiler will help us avoid anything leaking into the export templates (ie. if some non-editor code tries to access `is_runtime`, it'll be a compile-time error, because that property doesn't exist).

> Also, is the default value defined?

I'm not sure what you mean by this? The `p_extension_funcs` which is passed to us from the GDExtension needs to have all the fields on the struct filled out, and earlier in this function there's `bool is_runtime = (bool)p_extension_funcs->is_runtime`, so this value will definitely be set.

dsnopek commented on core/object/class_db.cpp:
```diff
@@ -70,6 +70,138 @@ HashMap<StringName, ClassDB::ClassInfo> ClassDB::classes;
+		return true;
+	}
+
+	static const GDExtensionPropertyInfo *placeholder_instance_get_property_list(GDExtensionClassInstancePtr p_instance, uint32_t *r_count) {
```
No, this function is used to get dynamic properties (it's `_get_property_list()`) and just like in GDScript we don't support dynamic properties on placeholders. The static properties are what's stored in `ClassDB`.

dsnopek commented on core/extension/gdextension.cpp:
```diff
@@ -1060,7 +1103,7 @@ void GDExtension::prepare_reload() {
 				state.push_back(Pair<String, Variant>(P.name, value));
 			}
-			E.value.instance_state[obj_id] = state;
+			E.value.instance_state[obj_id] = { state, obj->is_extension_placeholder() };
```
Designated initializers are a C++20 feature and I had thought we were still targeting C++17?

In any case, in other GDExtension code we've been putting the property names in inline comments, so in my latest push I've updated this code to be consistent with that :-)

dsnopek replied:
Yeah, C has had designated initializers for much longer, they're part of C99, but they were only added to C++ in C++20.

dsnopek commented:
@Riteo Thanks for the review! :heart:

deralmas reviewed (approved):
I haven't tested this but code wise looks great!

Bromeon commented on core/object/class_db.cpp:
```diff
@@ -70,6 +70,138 @@ HashMap<StringName, ClassDB::ClassInfo> ClassDB::classes;
+
+	static GDExtensionBool placeholder_instance_set(GDExtensionClassInstancePtr p_instance, GDExtensionConstStringNamePtr p_name, GDExtensionConstVariantPtr p_value) {
+		PlaceholderExtensionInstance *self = (PlaceholderExtensionInstance *)p_instance;
+		StringName name = *(StringName *)p_name;
```
This would probably not need to be copied (even if `StringName` copies are quite cheap)?

dsnopek replied:
In my latest push, I've change this (and a couple other places) to use const references to avoid the copy (at least until absolutely necessary).

Bromeon reviewed (commented):
Are there more naming changes to be expected, like the "tool/non-tool" discussion? I'd like to give this a try but each API change comes with quite a bit of overhead (fetch again, recompile Godot, generate bindings, adapt code) 🙂

dsnopek commented:
@Bromeon:

> Are there more naming changes to be expected, like the "tool/non-tool" discussion?

I really hope not :-)

akien-mga commented:
Thanks!
